src/sctools/metrics/metricsrunner.py

```diff
@@ -0,0 +1,61 @@
+import argparse
+import pysam
+
+
+class MetricsRunner:
+
+    def __init__(self, sam_file, open_mode=None):
+        """
+        :param str alignment_file: sam or bam file.
+        :param str open_mode: optional, mode to read file. Will be autodetected by file type if
+          the file contains the correct suffix for its type.
+        """
+
+        if open_mode is None:
+            if sam_file.endswith('.bam'):
+                open_mode = 'rb'
+            elif sam_file.endswith('.sam'):
+                open_mode = 'r'
+            else:
+                raise ValueError('could not autodetect file type for alignment_file %s '
+                                 '(detectible suffixes: .sam, .bam)' % sam_file)
+        self.input_sam = sam_file
+        self.open_mode = open_mode
+
+    def run_metrics(self, metrics_to_run):
+        """
+        :param list metrics_to_run: list of metrics to run for each record in input_sam
+        """
+
+        for metric in metrics_to_run:
+            metric.setup()
+
+        with pysam.AlignmentFile(self.input_sam, self.open_mode) as sam:
+            for record in sam:
+                for metric in metrics_to_run:
+                    metric.gather_metric(record)
+
+        for metric in metrics_to_run:
+            metric.calculate_and_output()
+
+
+def convert_class_name_to_class(metric_class_names):
+    # maybe some kind of case statement or python equivalent to grab a class from a string
+    # once we define a metric class that extends basemetric.py we can fill out this method
+    pass
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("input_bam", help="path to input bam file")
+    parser.add_argument("-m", "--metric", action="append", help="metrics to run on input_bam")
+    args = parser.parse_args()
+
+    metric_classes = None  # convert_class_name_to_class(args.metric)
```

Comment:
I think a neat way of doing this would be as follows (with better names for the directory structure): 

Make the structure something like this: 
```
metrics
__init__.py
runner.py
base_metric.py
classes/__init__.py
classes/metric_one.py
classes/metric_two.py
...
classes/metric_n.py
```

**`metrics/classes/__init__.py:`**
```
from .metric_one import metric_one
from .metric_two import metric_two
...
from .metric_n import metric_n
```
**`metrics/runner.py:`**
```
from metrics import classes
...

metrics = [getattr(classes, metric_name) for metric_name in metrics]
```

Reply:
definitely was not thinking about structure yet but I'm not sure if `/classes/` is the best sub dir name but I cant think of a better one so we'll stick with it for now!  I'm not even sure if `metrics` should be the root dir name like it currently is.  Always room for improvement.

You will probably have to explain to me how `from metrics import classes` works 😬 .

Reply:
It triggers the `classes/__init__.py` and dumps the results into the `classes` namespace, now in your global namespace.

Reply:
so `classes.metric_one` becomes addressable, which means that getattr(classes, "metric_one") will produce `metric_one`.